### DISCUSSION TITLE
Multiple manifests

### DISCUSSION THREAD

**Discussion by dcnieho** [microsoft/vcpkg#18698] [Q&A]
I have a solution laid out like this:
\    (root, contains base_project)
\dependencies
\project_using_base

where \dependencies contains the dependencies of both base_project and project_using_base (some shared, some unique. There are some dependency headers that project_using_base doesn't use itself, but are needed by the headers of base_project, and thus indirectly needed to build project_using_base.

I really like manifest mode, and was looking to use one manifest for each project, using \dependencies\vcpkg_installed as the place where both install their dependencies so i can easily set up my include directories for each project. That would mean one vcpkg.json in \ for base_project and one in \project_using_base for project_using_base.

However i run into these issues:
1. building one project removes all the dependencies of the other (using msbuild integration), and that means that when building sometimes required headers are missing. For example: base_project depends on a and b, and b is included in its headers and thus exposed through there. project_using_base depends on a and c, but not b. A solution build will first build base_project putting a and b in \dependencies\vcpkg_installed. Then project_using_base is built, and c is added to \dependencies\vcpkg_installed, but b is removed from it. That causes build to fail.
2. vcpkg is not always invoked: When triggering build on the project_using_base while a source file of base_project was modified, MSVC correctly detects that base_project needs to be rebuilt first. However, the vcpkg action isn't done, so base_project is missing its dependencies and build fails.

Am i misunderstanding how this is supposed to work (or even the definition of dependency)? Is the explicit intent that each project have its own vcpkg_installed directory? And if so, is it possible to have the manifest of project_using_base point to base_project and pull in all of base_project's dependencies as well, to avoid the issue i am having that project_using_base fails to build because base_project's dependencies are missing?

**Comment by timmeh87** [comment]
In my opinion if you really want to use two manifest files then you should be directing the output to two different folders instead of constantly fighting to overwrite it. Seems like a lot of unnecessary churn. The extra cost would be storing all the libraries on your hard drive twice but in practice I think storage space is cheaper than time (waiting for constant rebuilds of different manifests into the same destination). I believe that vcpkg guarantees that the binaries across two runs will be identical

I have a situation where more than one project will use vcpkg and they might also include each other, so I just have one giant manifest that gets every library I would ever need, and I build all my projects against the output of that one manifest. It lets me use the version pinning feature without worrying that there will be a mismatch between different manifests. I can do one vcpkg run to get all the libraries I need and then just work off that until I need to include a new library. I dont use integration at all but thats a personal choice.

This would be similar to the "original" behavior of vcpkg where you just "vcpkg install" all the packages you want and they reside in a single central location accessed by all projects. Except I am using a manifest so I get all the benefits of that.

**Comment by PhoebeHui** [comment]
```
A  -- \ (root, contains base_project) 
B -- \project_using_base
C -- \dependencies 

The dependency chain is
A -> C1, C2
B -> A, C1, C3

The dependency change to
A -> C1, C2
B ->C1, C2, C3

So B will fail when C1, C2 or C3 was removed, and A will fail when C1 or C2 was removed, this is expected.
```

It look you need to resolve the dependency for B->A, then when you using B, then C1, C2 and C3 should be installed.

We have a same discussion https://github.com/microsoft/vcpkg/discussions/18122 for multi manifest file.
